src/transformers/models/musicgen/modeling_musicgen.py

```diff
@@ -0,0 +1,2360 @@
+            # first fill the upper triangular part (the EOS padding)
+            delay_pattern = torch.triu(
+                torch.ones((num_codebooks, max_length), dtype=torch.bool), diagonal=max_length - num_codebooks + 1
+            )
+            # then fill the lower triangular part (the BOS padding)
+            delay_pattern = delay_pattern + torch.tril(torch.ones((num_codebooks, max_length), dtype=torch.bool))
+            mask = ~delay_pattern.to(input_ids.device)
+            input_ids = mask * input_ids_shifted + ~mask * pad_token_id
+
+        else:
+            # we have padded decoder input ids - put them at the start of our pattern matrix then build the delay
+            # pattern with an offset. This is a WIP
+            for batch in range(bsz):
+                # how many padding tokens have we got
+                offset = sum(~padding_mask[batch, 0].bool())
+                # put the masked tokens at the start
+                input_ids_shifted[batch, :, :offset] = input_ids[batch, :, -offset:]
+                for codebook in range(num_codebooks):
+                    input_ids_shifted[batch, codebook, offset : codebook + offset] = pad_token_id
+                    input_ids_shifted[batch, codebook, offset + codebook : seq_len + codebook - offset] = input_ids[
+                        batch, codebook, :offset
+                    ]
+
+            # fill the upper triangular portion of our matrix
+            delay_pattern = torch.triu(
+                torch.ones((num_codebooks, max_length), dtype=torch.bool), diagonal=max_length - num_codebooks + 1
+            )
+            mask = ~delay_pattern.to(input_ids.device)
+            input_ids = mask * input_ids_shifted + ~mask * pad_token_id
+
+        # find the first position to start generating - this is the first place we have the -1 token
+        # and will always be in the first codebook (since it has no codebook offset)
+        first_codebook_ids = input_ids[:, 0, :]
+        start_ids = (first_codebook_ids == -1).nonzero()[:, 1]
+        first_start_id = min(start_ids)
+
+        # (bsz * num_codebooks, seq_len) -> (bsz, num_codebooks, seq_len)
+        pattern_mask = input_ids.reshape(bsz * num_codebooks, -1)
+        input_ids = input_ids[..., :first_start_id].reshape(bsz * num_codebooks, -1)
+        return input_ids, pattern_mask
+
+    @staticmethod
+    def apply_delay_pattern_mask(input_ids, decoder_pad_token_mask):
+        """Apply a delay pattern mask to the decoder input ids, only preserving predictions where
+        the mask is set to -1, and otherwise setting to the value detailed in the mask."""
+        seq_len = input_ids.shape[-1]
+        decoder_pad_token_mask = decoder_pad_token_mask[..., :seq_len]
+        input_ids = torch.where(decoder_pad_token_mask == -1, input_ids, decoder_pad_token_mask)
+        return input_ids
+
+    @torch.no_grad()
+    def generate(
```

Comment:
Refactored the `.generate` method as discussed @patrickvonplaten - would be great to hear your thoughts on this before committing to it for the final design! Will update the docstrings as required once we've settled on the design.

Overall, my thoughts are that this approach to explicitly calling `greedy_search` and `sample` works. It's not super compact, but is definitely easier to understand than the `super().generate` call that we had before. It will allow us to build on top with other generation methods (e.g. assisted generation) in the future, so I think it's the way to go